core/rpc/client/function/jsonrpc/client.go

```diff
@@ -0,0 +1,64 @@
+// package jsonrpc implements a JSON-RPC client for the Kwil function service.
+
+package jsonrpc
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+
+	"github.com/kwilteam/kwil-db/core/crypto/auth"
+	rpcclient "github.com/kwilteam/kwil-db/core/rpc/client"
+	"github.com/kwilteam/kwil-db/core/rpc/client/function"
+	jsonFunction "github.com/kwilteam/kwil-db/core/rpc/json/function"
+)
+
+// Client is a JSON-RPC client for the Kwil function service. It uses the
+// JSONRPCClient from the rpcclient package for the actual JSON-RPC communication,
+// and implements function service methods.
+type Client struct {
+	*rpcclient.JSONRPCClient
```

Comment:
> I wrapped `JSONRPCClient` instead of `core/rpc/client/user.Client`, because I think the design we have for `core/rpc` package is only handle certain service's RPC, and then assemble them in `core/client` package.

That make sense, yes.
 
> I also saw how @jchappelow impl `core/rpc/client/admin`, which make a lot sense, I probably should implement function client just like that. One thing I'm not sure about this approach is, if we have more services in the future(highly possible imo), how to assemble them? Maybe we should register different methods to `JSONRPCClient`, just like how JSONRPC server register methods/handlers

In for example `core/rpc/client/user/jsonrpc`, could we supplement the current constructor, which is:

```go
func NewClient(url *url.URL, opts ...rpcclient.RPCClientOpts) *Client {
	return &Client{
		JSONRPCClient: rpcclient.NewJSONRPCClient(url, opts...),
	}
}
```

with this complementary constructor:

```go
func WrapBaseClient(url *url.URL, cl *rpcclient.JSONRPCClient) *Client {
	return &Client{
		JSONRPCClient: cl,
	}
}
```

So at `core/client` level we can compose like you suggest.  That is, some type at that higher level would make the `rpcclient.JSONRPCClient` that is shared among the different method implementations in `core/rpc/...`?

This the the pattern I was contemplating in `core/rpc/client/admin/jsonrpc/client.go`:

```go
// NewClient constructs a new admin Client.
func NewClient(u *url.URL, opts ...rpcclient.RPCClientOpts) *Client {
	// alt: jsonclient.NewBaseClient() ... WrapBaseClient() ...
```

Reply:
But I think what I did with admin service was just the path of least resistance, and probably before we separated the base client.  I'm thinking of the base client as a "caller" now.

Reply:
yeah this looks good, the solution is much easier than I thought haha